dynamiqs/mesolve/me_solver.py

```diff
@@ -1,28 +1,23 @@
-import functools
-
 from torch import Tensor
 
-from ..solvers.solver import Solver, depends_on_H
-from ..utils.solver_utils import kraus_map
+from ..solvers.solver import Solver
+from ..utils.solver_utils import cache, kraus_map
 
 
 class MESolver(Solver):
     def __init__(self, *args, jump_ops: Tensor):
         super().__init__(*args)
         self.jump_ops = jump_ops[None, ...]  # (1, len(jump_ops), n, n)
+        self.sum_no_jump = (
+            (self.jump_ops.adjoint() @ self.jump_ops).squeeze(0).sum(dim=0)
+        )  # (n, n)
 
-    @functools.cached_property
-    def sum_no_jump(self) -> Tensor:
-        # -> (n, n)
-        return (self.jump_ops.adjoint() @ self.jump_ops).squeeze(0).sum(dim=0)
-
-    @depends_on_H
-    def H_nh(self, t: float) -> Tensor:
-        # non-hermitian Hamiltonian at time t
-        # -> (b_H, 1, n, n)
-        return self.H(t) - 0.5j * self.sum_no_jump
+        # define cached operator
+        # non-hermitian Hamiltonian
+        self.H_nh = cache(lambda H: H - 0.5j * self.sum_no_jump)  # (b_H, 1, n, n)
 
     def lindbladian(self, t: float, rho: Tensor) -> Tensor:
-        H_nh_rho = self.H_nh(t) @ rho  # (b_H, b_rho, n, n)
+        H = self.H(t)
+        H_nh_rho = self.H_nh(H) @ rho  # (b_H, b_rho, n, n)
```

Comment:
Maybe group these two lines together? I don't know how important it is to not define new objects, but here it's free

Reply:
If it doesn't change anything, keep it as you did since it's more readable.

Reply:
Yes I thought it'd be more readable by splitting. Also for someone who wants to read along and understand how the caching works nothing is hidden: `H` is a new tensor only if `self.H(t)` returns something new, and the subsequent tensors are only re-computed if `H` is a new tensor.